File: kargerMinCut/java/src/main/java/KargerMinCut.java
import java.util.*;
import java.util.stream.Collectors;

public class KargerMinCut {

    public static int doMinCut(int[][] edges) {
        if (edges == null || edges.length == 0){
            return 0;
        }

        int vertexCount =caculateVertexCount(edges);
        int minCut = edges.length;
        for(int i = 0;i< vertexCount*vertexCount*vertexCount;i++) {
            int tmpMinCut = doKargerMinCut(edges);
            if(tmpMinCut < minCut) {
                minCut = tmpMinCut;
            }
        }
        return minCut;
    }


    private static int doKargerMinCut(int[][] edges) {

        List<List<Integer>> cloneEdges = Arrays.asList(edges).stream().map(edge ->
            Arrays.asList(Integer.valueOf(edge[0]), Integer.valueOf(edge[1]))
        ).collect(Collectors.toList());


        Map<Integer, Integer> vertexMergeIdxMap = new HashMap<>();
        Map<Integer, List<Integer>> mergeIdxVertexMap = new HashMap<>();
        int maxVertex = 0;

        for (List<Integer> edge: cloneEdges) {
            Integer v0 = Integer.valueOf(edge.get(0));
            Integer v1 = Integer.valueOf(edge.get(1));

            vertexMergeIdxMap.put(v0, v0);
            vertexMergeIdxMap.put(v1, v1);

            if(mergeIdxVertexMap.get(v0) == null) {
                mergeIdxVertexMap.put(v0, new LinkedList<>());
                mergeIdxVertexMap.get(v0).add(v0);
            }
            if(mergeIdxVertexMap.get(v1) == null) {
                mergeIdxVertexMap.put(v1, new LinkedList<>());
                mergeIdxVertexMap.get(v1).add(v1);
            }

            if(v0.intValue() > maxVertex) {
                maxVertex = v0.intValue();
            }
            if(v1.intValue() > maxVertex) {
                maxVertex = v1.intValue();
            }
        }

        Random r = new Random();
        r.setSeed(System.currentTimeMillis());
        int mergeCount = 0;
        while (mergeCount < vertexMergeIdxMap.size()-2) {
            mergeCount++;
            Integer newMergeIdx = Integer.valueOf(mergeCount + maxVertex);

            int pickIdx = r.nextInt(cloneEdges.size());
            List<Integer> pickEdge = cloneEdges.get(pickIdx);
            Integer v0 = Integer.valueOf(pickEdge.get(0));
            Integer mergeIdx0 = vertexMergeIdxMap.get(v0);
            Integer v1 = Integer.valueOf(pickEdge.get(1));
            Integer mergeIdx1 = vertexMergeIdxMap.get(v1);

            if (mergeIdx0.intValue() != mergeIdx1.intValue()) {
                for(Integer v: mergeIdxVertexMap.get(mergeIdx0)) {
                    vertexMergeIdxMap.put(v, newMergeIdx);
                }
                for(Integer v: mergeIdxVertexMap.get(mergeIdx1)) {
                    vertexMergeIdxMap.put(v, newMergeIdx);
                }

                mergeIdxVertexMap.get(mergeIdx0).addAll(mergeIdxVertexMap.get(mergeIdx1));
                mergeIdxVertexMap.put(newMergeIdx, mergeIdxVertexMap.get(mergeIdx0));

                mergeIdxVertexMap.remove(mergeIdx0);
                mergeIdxVertexMap.remove(mergeIdx1);


            }

            cloneEdges.remove(pickEdge);
        }

        int minCut = 0;

        for(List<Integer> edge: cloneEdges) {
            Integer v0 = Integer.valueOf(edge.get(0));
            Integer mergeIdx0 = vertexMergeIdxMap.get(v0);
            Integer v1 = Integer.valueOf(edge.get(1));
            Integer mergeIdx1 = vertexMergeIdxMap.get(v1);
            if (mergeIdx0.intValue() != mergeIdx1.intValue()) {
                minCut++;
            }
        }

        return minCut;
    }

    private static int caculateVertexCount(int[][] edges) {

        Map<Integer, Boolean> map = new HashMap<>();
        for (int[] edge: edges) {
            Integer v0 = Integer.valueOf(edge[0]);
            Integer v1 = Integer.valueOf(edge[1]);

            map.put(v0, Boolean.TRUE);
            map.put(v1, Boolean.TRUE);
        }

        return map.size();
    }
}

File: kargerMinCut/python3/karger_mincut.py
from random import randrange

def do_min_cut(edges):
    if edges == None or len(edges) == 0:
        return 0

    vertex_count = _count_vertex(edges)
    min_cut = len(edges)
    for i in range(vertex_count*vertex_count):
        tmp_min_cut = _do_karger_min_cut(edges)
        if tmp_min_cut < min_cut:
            min_cut = tmp_min_cut

    return min_cut

def _do_karger_min_cut(edges):
    clone_edges = edges.copy()

    max_vertex, vertex_mergeid_map, mergeid_vertex_map = _init_vertex_mergeid_map(edges)

    merge_count = 0
    while merge_count < len(vertex_mergeid_map) - 2:
        pick_idx = randrange(len(clone_edges))
        pick_edge = clone_edges[pick_idx]
        v0 = pick_edge[0]
        m0 = vertex_mergeid_map[v0]
        v1 = pick_edge[1]
        m1 = vertex_mergeid_map[v1]

        if m0 != m1:
            merge_count += 1
            cur_mergeid = merge_count + max_vertex

            for v in mergeid_vertex_map[m0]:
                vertex_mergeid_map[v] = cur_mergeid
            for v in mergeid_vertex_map[m1]:
                vertex_mergeid_map[v] = cur_mergeid

            merge_group0 = mergeid_vertex_map[m0]
            merge_group1 = mergeid_vertex_map[m1]
            merge_group0.extend(merge_group1)

            mergeid_vertex_map[cur_mergeid] = merge_group0
            
            del mergeid_vertex_map[m0]
            del mergeid_vertex_map[m1]

        del clone_edges[pick_idx]

    min_cut = 0
    for edge in clone_edges:
        v0 = edge[0]
        m0 = vertex_mergeid_map[v0]
        v1 = edge[1]
        m1 = vertex_mergeid_map[v1]
        if m0 != m1:
            min_cut+=1

    return min_cut



def _init_vertex_mergeid_map(edges):
    max_vertex = 0
    vertex_mergeid_map = {}
    mergeid_vertex_map = {}

    for edge in edges:
        v0 = edge[0]
        v1 = edge[1]

        vertex_mergeid_map[v0] = v0
        vertex_mergeid_map[v1] = v1

        mergeid_vertex_map[v0] = [v0]
        mergeid_vertex_map[v1] = [v1]

        if v0>max_vertex:
            max_vertex = v0
        if v1>max_vertex:
            max_vertex = v1
    return max_vertex, vertex_mergeid_map, mergeid_vertex_map

def _count_vertex(edges):
    vertex_map = {}
    for edge in edges:
        vertex_map[edge[0]] = True
        vertex_map[edge[1]] = True

    return len(vertex_map)